Add PWA installability support for the Elastic skin

andreiapostu opened the pull request:
Over the last few months I’ve been using Roundcube Webmail as my daily email client, both on desktop and mobile, and I’ve often wished it could be installable as a Progressive Web App. After looking into the topic, I found an older pull request ([#9354](https://github.com/roundcube/roundcubemail/pull/9354)) that attempted to introduce this feature. Looking into that PR, I realized that while the idea was excellent, the implementation had grown confusing and was never completed.

Because I believe this feature would bring real value to Roundcube users, I decided to start again from scratch and rebuild everything cleanly, keeping the scope small and focused.

---

### Icon design and rationale

I designed two new icons for the PWA, one at 512×512 and one at 192×192.
In the previous PR, there were dozens of icon sizes for multiple platforms, which made replacing or customizing the logo unnecessarily complicated. My goal was to simplify maintenance while still meeting modern PWA requirements, so I limited the icon set to the two key dimensions recognized by Chrome, Firefox, and Safari.

I am open to discussion if the team prefers to reduce the set to a single image or extend it further, but I personally believe two prepared sizes are enough to ensure compatibility and crisp rendering across devices.

Both icons are PNGs created by placing the existing Roundcube SVG logo at the center of a solid background using the color `#2f3a3f`.
This color was taken directly from the Elastic theme styles at [`/skins/elastic/styles/colors.less#L41`](https://github.com/roundcube/roundcubemail/blob/master/skins/elastic/styles/colors.less#L41),
so that the icons visually match the identity and friendly aesthetic of Roundcube itself.

Regarding layout and spacing, I kept horizontal padding around 14% of the icon’s total width, inspired by the design recommendations from [Android’s Adaptive Icon guidelines](https://developer.android.com/develop/ui/views/launch/icon_design_adaptive) and [Apple’s Human Interface Guidelines for app icons](https://developer.apple.com/design/human-interface-guidelines/app-icons). I am not a professional designer, but I tried to create something simple and coherent Roundcube elastic design language.

Below are previews of the icon I rendered, along with screenshots of how it appears once installed on iOS homescreen.
<img width="512" height="512" alt="icon-512" src="https://github.com/user-attachments/assets/c6755633-c57b-424e-b2d5-5766df3fc135" />
<img width="370" height="344" alt="ios_render" src="https://github.com/user-attachments/assets/acc8f187-0834-4dcb-9ad0-999a1f6bb0e9" />

---

### Manifest details

Here is the `manifest.json` file added under `skins/elastic/pwa/`:

```json
{
    "name": "Roundcube Webmail",
    "short_name": "Roundcube",
    "description": "Modern webmail for everyday use.",
    "start_url": "../../../",
    "scope": "../../../",
    "display": "standalone",
    "theme_color": "#2f3a3f",
    "background_color": "#2f3a3f",
    "icons": [
        {
            "src": "../images/icon-192.png",
            "sizes": "192x192",
            "type": "image/png"
        },
        {
            "src": "../images/icon-512.png",
            "sizes": "512x512",
            "type": "image/png"
        }
    ]
}
```

The `start_url` and `scope` fields may look unusual at first glance.
I chose `"../../../"` intentionally to ensure Roundcube remains installable regardless of its installation directory. Many users host Roundcube under paths such as `/mail/`, `/rc/`, or `/webmail/`, rather than at the root of their domain. Using relative paths ensures the manifest points to the application’s root and not to the skin’s internal folder. I’m open to suggestions on whether there might be a cleaner or more standardized way to achieve the same goal.

For `theme_color` and `background_color`, I used the same `#2f3a3f` tone mentioned above, again to stay consistent with the dark variant of the Elastic interface. However, this could be adjusted in the future to default to the light mode palette if the team prefers.

The manifest was placed in a new directory, `skins/elastic/pwa/`. For now, this directory only contains `manifest.json`, but it could later become the natural home for other PWA-related assets, such as a service worker.

---

### About the service worker

I’m fully aware that offline access is a key capability of any complete PWA.
For now, this pull request focuses solely on making Roundcube installable, nothing more. Writing an effective service worker for Roundcube would require a deep understanding of its caching layers, authentication flow, and internal asset loading. That task is complex and deserves a dedicated discussion and its own separate pull request. My intention would be to build upon this foundation to continue improving PWA capabilities in the future.

---

Making and thinking about a PWA is not as easy as it might seem.
There are many small technical and UX details to get right, and I’m trying to think carefully about each of them. I truly believe that having Roundcube as a proper PWA would significantly improve the user experience for many people, especially those who use it daily on mobile devices.

I’ve done my best to start this process with a clean, minimal, and well-reasoned implementation, and I would be glad to refine it further with feedback from the maintainers.

andreiapostu committed:
feat(elastic): add PWA manifest and base icons

andreiapostu committed:
feat(elastic): link PWA manifest in layout.html

andreiapostu committed:
feat(elastic): extend meta.json with mobile web app tags

pabzm commented on skins/elastic/meta.json:
```diff
@@ -15,6 +15,8 @@
-		"msapplication-navbutton-color": "#f4f4f4"
+		"msapplication-navbutton-color": "#f4f4f4",
+		"mobile-web-app-capable": "yes",
+		"application-name": "Roundcube"
```
Maybe we could introduce a `make` target that re-produces all relevant files from `$config['product_name']` and mention that in the comment for this config option?

pabzm replied:
That's right, but if someone does deployment from git, then I would expect them to be familiar with dealing with changed files. Don't you think so?

andreiapostu committed:
feat: dynamically inject application-name meta tag from config

andreiapostu committed:
feat(elastic): generate PWA manifest dynamically via data URI

andreiapostu commented:
Hi everyone, apologies for the long silence! I have been giving this a lot of thought.
I've updated the PR to address the concerns regarding `$config['product_name']`.

I've moved the `application-name` meta tag to be injected dynamically via PHP.

To avoid the issues with static files or git-ignored files mentioned by @alecpl and @pabzm, I've implemented a dynamic PWA manifest using a Data URI. This ensures the manifest always reflects the current `product_name` and `assets_path` without requiring any manual file generation or disk writes.

Regarding the suggestion to make this a plugin: while I see the benefit of decoupling, PWA support is increasingly a standard expectation for modern web interfaces. By keeping it integrated into the Elastic skin, we provide a seamless out-of-the-box experience for the majority of users.

Happy to hear your thoughts on this updated approach!

johndoh commented:
I still think this would be better as a plugin.  `meta.json` is not upgrade safe and even if it was the `mobile-web-app-capable` option which is added is not referenced anywhere. In addition neither the brand color or the description are configurable so this PR forces all admins to conform to a specific brand with no documentation changes. I worry it will add to the idea that Roundcube is a service provider rather than an app used by service providers because there is no way to disable or modify the the proposed functionally.

seb1k commented:
Hi !

Did u try the PWA from my skin : https://github.com/seb1k/Elastic2022
I need some feedback to know if everything is ok.

Thanks !
